File: 2nd Semester/Tree/Tree.cpp
#include <iostream>
#include <limits.h>
#include <cmath>
#include <list>
#include <string>

using namespace std;

struct TreeNode {
    int data;
    int layer;
    TreeNode *left;
    TreeNode *right;

    int getData() {
        return this->data;
    }
    bool hasLeft() {
        return this->left != nullptr;
    }
    bool hasRight() {
        return this->right != nullptr;  
    }
    int getLayer() {
        return this->layer;
    }
    void setLayer(int newLayer) {
        this->layer = newLayer;
    }
};

class Tree {
private:
    TreeNode* root;
    int size;
    list<TreeNode*> children;

public:
    Tree(int data) {
        this->root = new TreeNode;
        this->root->data = data;
        this->root->left = nullptr;
        this->root->right = nullptr;
        this->size = 1;
        this->root->layer = 0;
        this->children.push_back(this->root);
    }
    Tree() {
        this->root = nullptr;
        this->size = 0;
    }
    ~Tree() {};

    int getSize() {
        return this->size;
    }

    TreeNode* createNode(int data, TreeNode* l = nullptr, TreeNode* r = nullptr) {
        TreeNode* newNode = new TreeNode;
        newNode->data = data;
        newNode->left = l;
        newNode->right = r;
        newNode->layer = 0;
        return newNode;
    }

    void add(int data) {
        if (this->size == 0) {
            this->root = createNode(data);
            this->children.push_back(this->root);
        } else {
            for (auto node : this->children) {
                if (!node->hasLeft()) {
                    node->left = createNode(data);
                    node->left->layer = node->layer + 1;
                    this->children.push_back(node->left);
                    break;
                } else if (!node->hasRight()) {
                    node->right = createNode(data);
                    node->right->layer = node->layer + 1;
                    this->children.push_back(node->right);
                    break;
                }
            }
        }
        this->size += 1;
    }

    string displayLayer(int layer) {
        string s = "";
    
        // Add data of nodes in the layer
        for (auto node : this->children) {
            if (node->layer == layer) {
                s += to_string(node->getData());
                s += " ";
        }
        }
        return s;
    }

    bool search(int target) {
        return search(this->root, target);
    }

    bool search(TreeNode* node, int target) {
        if (node == nullptr) {
            return false;
        }

        if (node->getData() == target) {
            return true;
        }

        bool left = search(node->left, target);
        bool right = search(node->right, target);

        return left || right;
    }

    int countLeaves() {
        return countLeaves(this->root);
    }
    int countLeaves(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }
        if (!node->hasLeft() && !node->hasRight()) {
            return 1;
        }

        return countLeaves(node->left) + countLeaves(node->right);
    }

    Tree* copy() {
        return this;
    }

    void clear() {
        this->children.clear();
        this->root = NULL;
    }
    void printInOrder() {
        printInOrder(this->root);
    }
    void printInOrder(TreeNode* node) {
        if (node == nullptr) {
            return;
        }
        printInOrder(node->left);
        cout << node->getData() << " ";
        printInOrder(node->right);
    }

    bool hasPathSum(int sum) {
        return hasPathSum(this->root, sum);
    }
    bool hasPathSum(TreeNode* node, int sum) {
        if (node == nullptr) {
            return false;
        }
        if (node->getData() == sum && !node->hasLeft() && !node->hasRight()) {
            return true;
        }
        return hasPathSum(node->left, sum - node->getData()) || hasPathSum(node->right, sum - node->getData());
    }
};

int main() {
    Tree* t = new Tree(1);
    t->add(2);
    t->add(3);
    t->add(4);
    t->add(5);
    t->add(6);
    t->add(7);
    t->add(8);
    t->add(9);
    t->add(10);
    t->add(11);
    t->add(12);
    t->add(13);
    t->add(14);
    t->add(15);

    Tree* copy = t->copy();

    t->printInOrder();

    cout << "\n" << endl;

    //pathsum
    cout << "PathSum True: "  << t->hasPathSum(15) << endl;

    // should be false
    cout << "PathSum false: " << t->hasPathSum(122) << endl;

    t->clear();
    cout << "The tree has been cleared." << endl;
    cout << "Size: " << t->getSize() << endl;

    cout << t->getSize() << endl;
    cout << copy->getSize() << endl;

    cout << t->search(9) << endl;
    for(int i=0; i<4; i++){
        cout << t->displayLayer(i) << endl;
    }
    return 0;
}

File: 2nd Semester/Tree/Java/Tree.java
import java.util.*;

@SuppressWarnings("unchecked")
public class Tree<E> {
    public int size;
    public int layer;
    public TreeNode<E> root;
    private ArrayList<TreeNode> children;

    public Tree(E data) {
        this.root = new TreeNode<>(data);
        this.root.setLayer(0);
        this.children = new ArrayList<TreeNode>();
        this.children.add(this.root);
        this.size = 1;
        this.layer = 0;
    }

    public Tree() {
        this.root = null;
        this.children = new ArrayList<TreeNode>();
        this.size = 0;
        this.layer = 0;
    }

    public TreeNode<E> getRoot() {
        return this.root;
    }

    public int getSize() {
        return this.size;
    }

    public int getNodeLevel(TreeNode node) {
        int index = -1;
        // Find the index of the node with the given data
        for (int i = 0; i < this.children.size(); i++) {
            if (this.children.get(i).getData().equals(node.data)) {
                index = i;
                break;
            }
        }
        if (index == -1) {
            // Node with the given data not found in the tree
            return -1;
        }

        // Calculate the level of the node based on its index
        int level = 0;
        while (index >= getMaxNodes(level)) {
            level++;
        }
        return level;
    }

    public int getMaxNodes(int level) {
        int sum = 0;
        for (int i = 0; i < level + 1; i++) {
            sum = (int) (sum + Math.pow(2, i));
        }
        return sum;
    }

    public void add(E data) {

        if (this.children.size() == 0) {
            this.root = new TreeNode<>(data);
            this.root.setLayer(0);
            this.children.add(this.root);
        } else {
            TreeNode<E> parent = this.children.get((this.size - 1) / 2); // Get the parent node
            TreeNode<E> newNode = new TreeNode<>(data);
            if (!parent.hasLeft()) {
                parent.left = newNode;
            } else {
                parent.right = newNode;
            }
            newNode.setLayer(parent.getLayer() + 1);
            this.children.add(newNode);
        }

        this.size += 1;
    }

    public String displayLayer(int layer) {
        String s = "";
        for (TreeNode node : this.children) {
            if (this.getNodeLevel(node) == layer)
                s = s + node.getData() + " ";
        }
        return s;
    }

    public TreeNode createNode(E data) {
        TreeNode node = new TreeNode(data);
        return node;
    }

    public boolean search(E target) {
        return search(this.root, target);
    }

    private boolean search(TreeNode<E> node, E target) {
        if (node == null)
            return false;

        if (node.getData().equals(target))
            return true;

        // Recursively search in the left and right subtrees
        boolean foundLeft = search(node.left, target);
        boolean foundRight = search(node.right, target);

        // Return true if the target is found in either subtree
        return foundLeft || foundRight;
    }

    public int countLeaves() {
        return countLeaves(this.root);
    }

    private int countLeaves(TreeNode<E> node) {
        if (node == null)
            return 0;

        if (!node.hasLeft() && !node.hasRight())
            return 1;

        int leftLeaves = countLeaves(node.left);
        int rightLeaves = countLeaves(node.right);

        return leftLeaves + rightLeaves;
    }

    public Tree copy() {
        return this;
    }

    public void clear() {
        this.children.clear();
        this.root = null;
    }

    public boolean hasPathSum(int amount) {
        return false;
    }

    public void printInOrder() {
        printInOrder(this.root);
    }

    private void printInOrder(TreeNode<E> node) {
        if (node == null)
            return;

        printInOrder(node.left);
        System.out.print(node.getData() + " ");
        printInOrder(node.right);
    }

    public static void main(String[] args) {
        Tree t = new Tree("1");
        t.add("2");
        t.add("3");
        t.add("4");
        t.add("5");
        t.add("6");
        t.add("7");
        t.add("8");
        t.add("9");

        System.out.println("\nPrint the List: ");
        t.printInOrder();

        t.clear();

        System.out.println("\n\nCleared the list...");
        t.printInOrder();
    }

}